librz/arch/isa/msp430/msp430_disas.c

```diff
@@ -104,8 +89,7 @@ static void remove_second_operand(struct msp430_cmd *cmd) {
 }
 
 /* TODO: This code is messy, needs to be refactored. */
-static int decode_emulation(ut16 instr, struct msp430_cmd *cmd) {
-	int ret = -1;
+static void decode_emulation(ut16 instr, struct msp430_cmd *cmd) {
```

Comment:
The `TODO` above can be removed, right? Code looks petty clean to me.
But I would remove the `ut8 as, ad, src, dst, bw, opcode;` declaration.
Also add an `else` case which gives an error message (or whatever is appropriate) and fill `cmd` with invalid data.

Reply:
Well, it's defintely a bit repetitive and could use some abstraction into a more declarative form.

Every condition has the same general form ` if (opcode == ... && src == ... [&& dst == ...])`, so I'm thinking of abstracting this into an array that declaratively specifies those tuples of values (so `{ {val1, val2, val3}, {val4, val5, val6}, ...}`) and then a single loop that will interpret each tuple in the array with the same if condition. Worth it or too much of an overengineering gymnastics ?

Also, this condition **should** have no `else` clause, it's totally fine if none of the conditions match, that just means the instruction isn't an emulated instructions. We can't know beforehand if the instruction is emulated, the MSP430 spec just lists a bunch of usage patterns of 2-operand instructions and says they're emulations for specific instructions, those are the patterns checked by this unusually large `if` statement. If none of the conditinos match, then the instruction isn't an emulated instruction.

Reply:
> this condition should have no else clause

Ok, wasn't sure if it was supposed to be like this.

Regarding cleaness: IMO this is fine. No need to overengineer here.